src/workerd/server/server.c++

```diff
@@ -913,6 +915,154 @@ kj::Own<Server::Service> Server::makeDiskDirectoryService(
+              // Nothing to do here but move along.
+              KJ_LOG(INFO, kj::str("Inspector client detached [", id, "]"));
+              return kj::READY_NOW;
+            } else {
+              // If it's any other kind of error, propagate it!
+              throw ex;
+            }
+          });
+        }
+
+        return response.sendError(404, "Unknown worker session", responseHeaders);
+      }
+
+      // No / in url!? That's weird
+      return response.sendError(400, "Invalid request", responseHeaders);
+    }
+
+    // If the request is not a WebSocket request, it must be a GET to fetch details
+    // about the implementation.
+    if (method != kj::HttpMethod::GET) {
+      return response.sendError(501, "Unsupported Operation", responseHeaders);
+    }
+
+    if (url.endsWith("/json/version")) {
+      responseHeaders.set(kj::HttpHeaderId::CONTENT_TYPE, "application/json"_kj);
+      auto content = kj::str("{\"Browser\": \"workerd\", \"Protocol-Version\": \"1.3\" }");
+      auto out = response.send(200, "OK", responseHeaders, content.size());
+      return out->write(content.begin(), content.size()).attach(kj::mv(content), kj::mv(out));
+    } else if (url.endsWith("/json") || url.endsWith("/json/list")) {
+      responseHeaders.set(kj::HttpHeaderId::CONTENT_TYPE, "application/json"_kj);
+
+      auto baseWsUrl = KJ_UNWRAP_OR(headers.get(kj::HttpHeaderId::HOST), {
+        return response.sendError(400, "Bad Request", responseHeaders);
+      });
+
+      kj::Vector<kj::String> entries(isolates.size());
+      for (auto& entry : isolates) {
+        kj::Vector<kj::String> fields(9);
+        fields.add(kj::str("\"id\":\"", entry.key ,"\""));
+        fields.add(kj::str("\"title\":\"workerd: worker ", entry.value.name ,"\""));
+        fields.add(kj::str("\"type\":\"node\""));
+        fields.add(kj::str("\"description\":\"workerd worker\""));
+        fields.add(kj::str("\"webSocketDebuggerUrl\":\"ws://",
+                            baseWsUrl ,"/", entry.key ,"\""));
+        fields.add(kj::str("\"devtoolsFrontendUrl\":\"devtools://devtools/bundled/js_app.html?experiments=true&v8only=true&ws=", baseWsUrl ,"/\""));
+        fields.add(kj::str("\"devtoolsFrontendUrlCompat\":\"devtools://devtools/bundled/inspector.html?experiments=true&v8only=true&ws=", baseWsUrl ,"/\""));
+        fields.add(kj::str("\"faviconUrl\":\"https://workers.cloudflare.com/favicon.ico\""));
+        fields.add(kj::str("\"url\":\"https://workers.dev\""));
+        entries.add(kj::str('{', kj::strArray(fields, ",") ,'}'));
+      }
+      auto content = kj::str('[', kj::strArray(entries, ","), ']');
+
+      auto out = response.send(200, "OK", responseHeaders, content.size());
+      return out->write(content.begin(), content.size()).attach(kj::mv(content), kj::mv(out));
+    }
+
+    return response.sendError(500, "Not yet implemented", responseHeaders);
+  }
+
+  void addIsolate(Worker::Isolate* isolate, kj::StringPtr name) {
```

Comment:
Hmm, it's a little weird that this permanently stores a pointer to `isolate`. But at present it's true that isolates are never destroyed so I guess it's fine. If we wanted to account for the possibility of shutting down isolates then this should return some sort of a handle object that can be dropped to un-register the isolate I suppose. But I won't make you do that now unless you feel like it. Otherwise there should probably be a TODO comment, though.

Reply:
yeah, this felt a bit icky to me also and I was thinking about revisiting it. The handle object is a good idea. Since I have a few other updates here I'll go ahead and make this change also.

Reply:
Ok, reworked this piece into the handle object.